### PR TITLE
HPCC-22941 Remember user's selection of rows

### PR DESCRIPTION
**Pull request by miguelvazq**
Initial implementation using ws_store to remember a users selection of
rows within a data table (dgrid)

Signed-off by: Miguel Vazquez <miguel.vazquez@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [x] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by miguelvazq** [comment]
@GordonSmith Tested with and without username to test the fall back mechanism of using local storage. Please review.

File: esp/src/eclwatch/DFUQueryWidget.js [hunk 1]
```diff
@@ -713,7 +713,7 @@ define([
                         context.downloadToList.set("disabled", true);
                     }
                 });
-                ESPUtil.goToPageUserPreference(this.workunitsGrid, "DFUQueryWidget");
+                ESPUtil.goToPageUserPreference(this.workunitsGrid, "DFUQueryWidget_GridRowsPerPage");
                 this.workunitsGrid.startup();
```
Comment (GordonSmith) [inline esp/src/eclwatch/DFUQueryWidget.js]: Are we sure these are actually set _before_ the grid startup happens?

File: esp/src/eclwatch/DynamicESDLQueryWidget.js [hunk 2]
```diff
@@ -192,6 +192,7 @@ define([
                     context.deleteButton.set("disabled", true);
                 }
             });
+            ESPUtil.goToPageUserPreference(retVal, "DynamicESDLQueryWidget_GridRowsPerPage");
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLQueryWidget.js]: Are we sure these are actually set _before_ the grid startup happens?

File: esp/src/eclwatch/EventScheduleWorkunitWidget.js [hunk 3]
```diff
@@ -165,6 +165,7 @@ define([
                 this.eventGrid.onSelectionChanged(function (event) {
                     context.refreshActionState();
                 });
+                ESPUtil.goToPageUserPreference(this.eventGrid, "EventScheduleWorkunitWidget_GridRowsPerPage");
```
Comment (GordonSmith) [inline esp/src/eclwatch/EventScheduleWorkunitWidget.js]: Are we sure these are actually set _before_ the grid startup happens?

File: esp/src/eclwatch/GetDFUWorkunitsWidget.js [hunk 4]
```diff
@@ -451,7 +451,7 @@ define([
                         context.downloadToList.set("disabled", true);
                     }
                 });
-                ESPUtil.goToPageUserPreference(this.workunitsGrid, "GetDFUWorkunitsWidget");
+                ESPUtil.goToPageUserPreference(this.workunitsGrid, "GetDFUWorkunitsWidget_GridRowsPerPage");
```
Comment (GordonSmith) [inline esp/src/eclwatch/GetDFUWorkunitsWidget.js]: Are we sure these are actually set _before_ the grid startup happens?

File: esp/src/eclwatch/QuerySetQueryWidget.js [hunk 5]
```diff
@@ -461,7 +461,7 @@ define([
                         context.downloadToList.set("disabled", true);
                     }
                 });
-                ESPUtil.goToPageUserPreference(this.querySetGrid, "QuerySetQueryWidget");
+                ESPUtil.goToPageUserPreference(this.querySetGrid, "QuerySetQueryWidget_GridRowsPerPage");
```
Comment (GordonSmith) [inline esp/src/eclwatch/QuerySetQueryWidget.js]: Are we sure these are actually set _before_ the grid startup happens?

File: esp/src/eclwatch/UserQueryWidget.js [hunk 6]
```diff
@@ -542,6 +542,7 @@ define([
             this.groupsGrid.onSelectionChanged(function (event) {
                 context.refreshActionState();
             });
+            ESPUtil.goToPageUserPreference(this.groupsGrid, "UsersQueryWidget_GroupsGrid_GridRowsPerPage");
```
Comment (GordonSmith) [inline esp/src/eclwatch/UserQueryWidget.js]: Are we sure these are actually set _before_ the grid startup happens?

File: esp/src/eclwatch/UserQueryWidget.js [hunk 7]
```diff
@@ -655,6 +656,7 @@ define([
             this.usersGrid.onSelectionChanged(function (event) {
                 context.refreshActionState();
             });
+            ESPUtil.goToPageUserPreference(this.usersGrid, "UsersQueryWidget_UsersGrid_GridRowsPerPage");
```
Comment (GordonSmith) [inline esp/src/eclwatch/UserQueryWidget.js]: Are we sure these are actually set _before_ the grid startup happens?

**Review by GordonSmith — COMMENTED** [review]
Code looks good - but suspect the startup for all the grids should be happening within the then callback?

File: esp/src/eclwatch/DFUQueryWidget.js (hunk 1 above)
Comment (GordonSmith) [reply esp/src/eclwatch/DFUQueryWidget.js]: Yes - but retVal is a `Promise` that something is going to happen in the future...

Comment (miguelvazq) [reply esp/src/eclwatch/DFUQueryWidget.js]: We can wrap it in a `then` to overcome any issues?

**Comment by miguelvazq** [comment]
@GordonSmith Updated.

File: esp/src/eclwatch/XrefQueryWidget.js [hunk 8]
```diff
@@ -127,6 +127,7 @@ define([
                     }
                 });
 
+                ESPUtil.goToPageUserPreference(retVal, "XrefQueryWidget_GridRowsPerPage");
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefQueryWidget.js]: This one may need testing...

File: esp/src/src/ESPUtil.ts [hunk 9]
```diff
@@ -38,6 +38,8 @@ import { select as d3Select } from "d3-selection";
 import { Pagination } from "./Pagination";
 
 import { declareDecorator } from './DeclareDecorator';
+// @ts-ignore
```
Comment (GordonSmith) [inline esp/src/src/ESPUtil.ts]: Why is the ignore needed?

**Review by GordonSmith — COMMENTED** [review]
Code looks ok - two questions.

File: esp/src/src/ESPUtil.ts (hunk 9 above)
Comment (miguelvazq) [reply esp/src/src/ESPUtil.ts]: No need removing.

**Comment by miguelvazq** [comment]
@GordonSmith bump.

File: esp/src/eclwatch/XrefQueryWidget.js (hunk 8 above)
Comment (GordonSmith) [reply esp/src/eclwatch/XrefQueryWidget.js]: You need to check where its being called from (as its in a function) and possibly return a promise from this function?

File: esp/src/eclwatch/DynamicESDLQueryWidget.js [hunk 10]
```diff
@@ -172,27 +172,28 @@ define([
...
-                    var item = retVal.row(evt).data;
-                    context._onRowDblClick(item);
-                }
-            });
+            ESPUtil.goToPageUserPreference(retVal, "DynamicESDLQueryWidget_GridRowsPerPage").then(function(){
+                retVal.on(".dgrid-row-url:click", function (evt) {
+                    if (context._onRowDblClick) {
+                        var item = retVal.row(evt).data;
+                        context._onRowDblClick(item);
+                    }
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLQueryWidget.js]: There is no need to put the "on" calls in the "then" - its that startup thats important.

File: esp/src/eclwatch/DynamicESDLQueryWidget.js [hunk 11]
```diff
@@ -172,27 +172,28 @@ define([
...
-                    context.deleteButton.set("disabled", true);
-                }
+                retVal.onSelectionChanged(function (event) {
+                    var selection = retVal.getSelected();
+                    if (selection.length > 0) {
+                        context.deleteButton.set("disabled", false);
+                    } else {
+                        context.deleteButton.set("disabled", true);
+                    }
+                });
```
Comment (GordonSmith) [inline esp/src/eclwatch/DynamicESDLQueryWidget.js]: There is no need to put the "on" calls in the "then" - its that startup thats important.

File: esp/src/eclwatch/WUQueryWidget.js [hunk 12]
```diff
@@ -516,13 +516,14 @@ define([
...
-                ESPUtil.goToPageUserPreference(this.workunitsGrid, "WUQueryWidget");
-                aspect.after(this.workunitsGrid, "gotoPage", function (deferred, args) {
-                    return deferred.then(function () {
-                        args[0] > 1 ? context._idleWatcher.stop() : context._idleWatcher.start();
+                ESPUtil.goToPageUserPreference(this.workunitsGrid, "WUQueryWidget_GridRowsPerPage").then(function(){
+                    aspect.after(context.workunitsGrid, "gotoPage", function (deferred, args) {
+                        return deferred.then(function () {
+                            args[0] > 1 ? context._idleWatcher.stop() : context._idleWatcher.start();
+                        });
                     });
```
Comment (GordonSmith) [inline esp/src/eclwatch/WUQueryWidget.js]: There is no need to put the "aspect.after" calls in the "then" - its that startup thats important.

File: esp/src/eclwatch/XrefQueryWidget.js [hunk 13]
```diff
@@ -114,19 +114,20 @@ define([
...
+                            var item = retVal.row(evt).data;
+                            context._onRowDblClick(item);
+                        }
+                    });
+                    retVal.on(".dgrid-row:dblclick", function (evt) {
+                        if (context._onRowDblClick) {
+                            var item = retVal.row(evt).data;
+                            context._onRowDblClick(item);
+                        }
+                    });
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefQueryWidget.js]: There is no need to put the "on" calls in the "then" - its that startup that is important.

File: esp/src/eclwatch/XrefQueryWidget.js [hunk 14]
```diff
@@ -114,19 +114,20 @@ define([
...
                 });
-
                 return retVal;
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefQueryWidget.js]: You need to check where this is being called from...

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Comment by miguelvazq** [comment]
@GordonSmith Updated. Please re-review.

File: esp/src/eclwatch/WUQueryWidget.js [hunk 15]
```diff
@@ -516,13 +516,15 @@ define([
                         context.downloadToList.set("disabled", true);
                     }
                 });
-                ESPUtil.goToPageUserPreference(this.workunitsGrid, "WUQueryWidget");
-                aspect.after(this.workunitsGrid, "gotoPage", function (deferred, args) {
+                ESPUtil.goToPageUserPreference(this.workunitsGrid, "WUQueryWidget_GridRowsPerPage").then(function(){
+                    context.workunitsGrid.startup();
+                });
+
+                aspect.after(context.workunitsGrid, "gotoPage", function (deferred, args) {
```
Comment (GordonSmith) [inline esp/src/eclwatch/WUQueryWidget.js]: Move the aspect to before the goToPageUserPreference call.

**Review by GordonSmith — CHANGES_REQUESTED** [review]
Just one minor change request.

File: esp/src/eclwatch/WUQueryWidget.js (hunk 15 above)
Comment (miguelvazq) [reply esp/src/eclwatch/WUQueryWidget.js]: Updated.

**Review by GordonSmith — APPROVED** [review]
Looks good...

**Comment by miguelvazq** [comment]
@richardkchapman Ready to merge.

**Comment by ghalliday** [comment]
@miguelvazq just checking if this was intended to go into 7.6.

**Comment by miguelvazq** [comment]
This is a new feature so it can go into 7.8. Can also wait until @richardkchapman merges.

**Pull request merged**